### ISSUE TITLE
Can't start on macOS

### ISSUE THREAD

**Issue by sj26** [machinio/cuprite#22]
Trying this out on macOS with `BROWSER_PATH=/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome` I get:

```
     1.1) Failure/Error: visit "/"

          URI::InvalidURIError:
            bad URI(is not URI?):
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/2.5.0/uri/rfc3986_parser.rb:18:in `rescue in split'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/2.5.0/uri/rfc3986_parser.rb:15:in `split'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/2.5.0/uri/rfc3986_parser.rb:73:in `parse'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/2.5.0/uri/common.rb:237:in `parse'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser/web_socket.rb:21:in `initialize'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser/client.rb:16:in `new'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser/client.rb:16:in `initialize'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser.rb:255:in `new'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser.rb:255:in `start'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser.rb:48:in `initialize'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser.rb:22:in `new'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/browser.rb:22:in `start'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/driver.rb:26:in `browser'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/cuprite-0.2.1/lib/capybara/cuprite/driver.rb:31:in `visit'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/capybara-3.9.0/lib/capybara/session.rb:265:in `visit'
          # /usr/local/rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/capybara-3.9.0/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
...
```

I popped open the gem and made Process.start dump output while looking for websockets and found:

```
[0118/045325.644206:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/045325.644206:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/045325.646487:ERROR:file_io.cc(89)] ReadExactly: expected 8, observed 0
[0118/045325.703505:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/045325.703505:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/045325.705365:ERROR:file_io.cc(89)] ReadExactly: expected 8, observed 0
```

Would be nice if the browser output were piped to the logger when `CUPRITE_DEBUG=1` or something.

**Comment by sj26** [comment]
If I start a process manually it starts fine:

```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --headless --disable-gpu --hide-scrollbars --mute-audio --enable-automation --disable-web-security --window-size=1024,768 --remote-debugging-port=0 --remote-debugging-address=127.0.0.1
[0118/155835.406981:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/155835.408449:ERROR:file_io.cc(89)] ReadExactly: expected 8, observed 0
[0118/155835.408905:ERROR:xattr.cc(64)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/sv/kt7w98jx5zj16c4whzrwmx9m0000gn/T/: Operation not permitted (1)
[0118/155835.451772:WARNING:dns_config_service_posix.cc(182)] dns_config has unhandled options!
[0118/155835.452613:ERROR:gpu_process_transport_factory.cc(967)] Lost UI shared context.

DevTools listening on ws://127.0.0.1:50152/devtools/browser/bc758536-c437-49af-8f04-359421e57c5d
```

I think it's just aborting too early

**Comment by sj26** [comment]
Confirmed, changing `attempts` from `3` to `10` gets it working

**Comment by sj26** [comment]
Making this based on elapsed cpu time seems sensible. This works well for me:

```
      def parse_ws_url(read_io)
        output = ""
        start = ::Process.clock_gettime(::Process::CLOCK_PROCESS_CPUTIME_ID)
        timeout = start + 1.0
        regexp = /DevTools listening on (ws:\/\/.*)/
        while (now = ::Process.clock_gettime(::Process::CLOCK_PROCESS_CPUTIME_ID)) < timeout
          begin
            output += read_io.read_nonblock(512)
          rescue IO::WaitReadable
            IO.select([read_io], nil, nil, timeout - now)
          else
            if output.match?(regexp)
              @ws_url = Addressable::URI.parse(output.match(regexp)[1])
              @host = @ws_url.host
              @port = @ws_url.port
              break
            end
          end
        end

        unless @ws_url
          raise "Chrome process did not produce websocket url within 1 second"
        end
      end
```

**Comment by route** [comment]
I decided to drop everything Chrome produces, didn't find anything important there except ws_url.
I like your solution, do you want to open a PR?

**Comment by iainbeeston** [comment]
For what it's worth I'm seeing this as well. Both solutions (increasing attempts to 10 and @sj26 's while loop) work on my project.

**Comment by iainbeeston** [comment]
I hope you don't mind @sj26 but I've raised a PR using your code, as I'd like to use it in my own project, and I figured if I was editing the code anyway I might as well raise it as a PR. I've credited you in the commit message. Let me know if you object and I'll close it right away.

FYI I've updated it slightly to use a monotonic clock (rather than CPU time) as suggested in [this](https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/) blog post.
